src/transport/mqtt/connection.rs

```diff
@@ -0,0 +1,760 @@
+    ///
+    /// Return [`None`] when disconnected.
+    pub(crate) async fn next_publish<S>(
+        &mut self,
+        client_id: ClientId<&str>,
+        interfaces: &Interfaces,
+        store: &StoreWrapper<S>,
+    ) -> Result<Option<Publish>, StoreError>
+    where
+        S: PropertyStore,
+    {
+        // While connected poll the connection for a publish event
+        while self.state.is_connected() {
+            let publish = self
+                .state
+                .poll(&mut self.connection, client_id, interfaces, store)
+                .await?;
+
+            if let Some(publish) = publish {
+                return Ok(Some(publish));
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Wait for the connection to be established to the Astarte.
+    ///
+    /// This function assumes that the connection is valid, but will handle disconnection and
+    /// reconnection automatically.
+    pub(crate) async fn wait_connack<S>(
+        client: AsyncClient,
+        eventloop: EventLoop,
+        provider: TransportProvider,
+        client_id: ClientId<&str>,
+        interfaces: &Interfaces,
+        store: &StoreWrapper<S>,
+    ) -> Result<Self, StoreError>
+    where
+        S: PropertyStore,
+    {
+        let mut connection = Connection {
+            client,
+            eventloop: SyncWrapper::new(eventloop),
+            provider,
+        };
+
+        let mut state = State::Connecting(Connecting);
+
+        // Wait till we are in the Init state, so we do not need to handle the incoming publishes,
+        // but all the initialization packets are being queued.
+        while !state.is_init() {
+            state
+                .poll(&mut connection, client_id, interfaces, store)
+                .await?;
+        }
+
+        Ok(Self { connection, state })
+    }
+}
+
+/// Struct to hold the connection and client to be passed to the state.
+///
+/// We pass a mutable reference to the connection from outside the state so we can freely move from
+/// one state to the other, returning a new one. This without having to move the connection from
+/// behind the mutable reference.
+#[derive(Debug)]
+struct Connection {
+    client: AsyncClient,
+    // NOTE: this should be replaces by Exclusive<EventLoop> when the feature `exclusive_wrapper`
+    //       is stabilized or the EventLoop becomes Sync
+    //       https://doc.rust-lang.org/std/sync/struct.Exclusive.html
+    eventloop: SyncWrapper<EventLoop>,
+    provider: TransportProvider,
+}
+
+impl Connection {
+    fn eventloop_mut(&mut self) -> &mut EventLoop {
+        self.eventloop.get_mut()
+    }
+}
+
+/// This cannot be a type state machine, because any additional data cannot be moved out of the enum
+/// when polling. The only data that can be easily changed is the current state into the next one.
+#[derive(Debug)]
+enum State {
+    Disconnected(Disconnected),
+    Connecting(Connecting),
+    StartInit(StartInit),
```

Comment:
What do you think to find a better name for `StartInit` state?
- ProtocolHandshake
- AstarteHandshake
- AstarteMQTTInit
- AstarteInit
- ...

Reply:
Yes, something like `Handshake` is more explanatory.